### PR TITLE
Usability improvements for timeout support in IndexSearcher

### LINKED ISSUE #11874: Usability improvements for timeout support in IndexSearcher
### Description

In OpenSearch, we used to rely on custom implementation for query timeouts support. Since `9.3`, Apache Lucene offers the timeout support in the `IndexSearcher` [1] but the implementation is quite restrictive. In the `OpenSearch` we would like to benefit from this new feature preferably without duplicating the code. 

The suggested usability improvements are low risk and do not open up any internals:
 - add getter for timeout value to `IndexSearcher` (only setter exists)
 - open up `TimeLimitingBulkScorer` and `TimeLimitingBulkScorer.TimeExceededException`

@msokolov would apprecite your opinion, thank you.

[1] https://issues.apache.org/jira/browse/LUCENE-10151

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

### Description

The suggested usability improvements are low risk and do not open up any internals:

 - add getter for timeout value to IndexSearcher (only setter exists)
 - open up TimeLimitingBulkScorer and TimeLimitingBulkScorer.TimeExceededException

Closes https://github.com/apache/lucene/issues/11874

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/search/TimeLimitingBulkScorer.java [hunk 1]
```diff
@@ -28,14 +28,14 @@
  *
  * @see org.apache.lucene.index.ExitableDirectoryReader
  */
-final class TimeLimitingBulkScorer extends BulkScorer {
+public final class TimeLimitingBulkScorer extends BulkScorer {
```
Comment (reta) [inline lucene/core/src/java/org/apache/lucene/search/TimeLimitingBulkScorer.java:31]: It used to be `public` but the visibility was reduced later on

File: lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java [hunk 2]
```diff
@@ -519,10 +524,15 @@ public void search(Query query, Collector results) throws IOException {
     search(leafContexts, createWeight(query, results.scoreMode(), 1), results);
   }
 
-  /** Returns true if any search hit the {@link #setTimeout(QueryTimeout) timeout}. */
+  /** Return true if any search hit the {@link #setTimeout(QueryTimeout) timeout}. */
```
Comment (reta) [inline lucene/core/src/java/org/apache/lucene/search/IndexSearcher.java:527]: Not related, fixing comment to keep it concise

**Comment by jpountz** [comment]
Adding a getter works for me, but I'd prefer not to make other implementation details like the custom bulk scorer public, why do you need this?

**Comment by reta** [comment]
Thanks @jpountz 

> Adding a getter works for me, 

:+1: 

> but I'd prefer not to make other implementation details like the custom bulk scorer public, why do you need this?

We do override `IndexSearcher::search` [1] and basically take care of  `BulkScorer` creation, at this point we could not wrap it up with time limiting one. The TimeLimitingBulkScorer is essentially a simple wrapper, to me there is nothing trully implementation specific / intrisic here (obviusly subjective opionion). Check for example TimeLimitingCollector / TimeLimitingCollector.TimeExceededException pair as an argument towards opening the bulk scorer, it is public API. 


[1] https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L276

**Comment by vigyasharma** [comment]
> but I'd prefer not to make other implementation details like the custom bulk scorer public, why do you need this?

I think they want to use they want to wrap their custom BulkScorer inside the `TimeLimitingBulkScorer`, and need access to it inside their overridden `IndexSearcher.search(...)` method. That would let them move away from their current [CancellableBulkScorer](https://github.com/opensearch-project/OpenSearch/blob/171d14153c8542f25b52428863f1ec9372efd671/server/src/main/java/org/opensearch/search/internal/CancellableBulkScorer.java#L48) impl. Since they override the search() method, their BulkScorer does not get wrapped automatically.

To me, it seems like `TimeLimitingBulkScorer` specifically, might be a useful class to expose outside lucene. @jpountz - can you elaborate more on the downsides of exposing this?

**Review by vigyasharma — APPROVED** [review]
Looks good to me. I'll wait for a few days before merging, in case people have comments/concerns with the public visibility for `TimeLimitingBulkScorer`

**Comment by reta** [comment]
> Looks good to me. I'll wait for a few days before merging, in case people have comments/concerns with the public visibility for `TimeLimitingBulkScorer`

Thanks a lot @vigyasharma !

**Comment by jpountz** [comment]
Sorry for the lag I'm on vacation.

The problem with "this class may be useful outside of Lucene" to me is that it could apply to any class in Lucene. We did indeed make some classes public on this criterion in the past (e.g FixedBitSet), but I would like the bar to be high, is there really not a better way? Is the need to make this class public a sign that the functionality is not exposed the right way?

For the case described above by Vigya, I can think of two alternatives we might want to consider instead:
 - Copy the code for this time-limiting bulk scorer instead of using the Lucene class. If there is only a couple users of Lucene who would benefit from this class being public, maybe it's a better trade-off to let them take full ownership of this code to allow Lucene to keep treating it as an implementation detail?
 - Figure out different refactors that would help OpenSearch leverage timeout support from IndexSearcher as-is. E.g. if it would fold the logic to use live docs to lead iteration in the weight wrapper, would that work?

**Comment by reta** [comment]
Thanks @jpountz, no problem at all.

> Sorry for the lag I'm on vacation.
> 
> The problem with "this class may be useful outside of Lucene" to me is that it could apply to any class in Lucene. We did indeed make some classes public on this criterion in the past (e.g FixedBitSet), but I would like the bar to be high, is there really not a better way? Is the need to make this class public a sign that the functionality is not exposed the right way?

I think you are getting straight feedback from the users (OpenSearch in this case, possibly Elasticsearch): this class is useful outside of Lucene for the reasons highlighted above. So from my perspective, it falls into "Is the need to make this class public a sign that the functionality is not exposed the right way?" bucket.

> 
> For the case described above by Vigya, I can think of two alternatives we might want to consider instead:
> 
> Copy the code for this time-limiting bulk scorer instead of using the Lucene class. If there is only a couple users of Lucene who would benefit from this class being public, maybe it's a better trade-off to let them take full ownership of this code to allow Lucene to keep treating it as an implementation detail?

Copy / Paste is always the option. I opened the issue to highlight that probably we could avoid that by addressing "functionality is not exposed the right way". From the contributor perspective, I see that historically a lot of things were copy / pasted from Lucene into OpenSearch for (probably) same reasons - it is simpler than suggesting the change.

> Figure out different refactors that would help OpenSearch leverage timeout support from IndexSearcher as-is. E.g. if it would fold the logic to use live docs to lead iteration in the weight wrapper, would that work?

The refactor is on the table for sure, the implementation is very different from the Lucene for reasons I have no knowledge of, sadly (at least at this moment), so this is not on the table.

**Comment by jpountz** [comment]
Historically configuring timeouts on searches has been too complicated: users had to wrap their collector with a `TimeLimitingCollector` and to wrap their readers with an `ExitableDirectoryReader` in order to make sure that other phases of searches like rewrites, collecting matching points into a bitset or vector searches would also honor the timeout. And I would bet that many users don't even know that they need to do both for timeouts to be applied correctly.

So in #927 we added a more user-friendly `IndexSearcher#setTimetout` with the idea that it would do the right thing automatically. For now it only supports aborting searches at collection time, but there are open follow-ups to make it better:
 - #11676
 - #11677
 - #11700

When these issues are addressed, I'm hoping to deprecate `TimeLimitingCollector` and update javadocs of `ExitableDirectoryReader` to make it clear that it's a very expert class that is only useful when consuming index readers directly as opposed to through an `IndexSearcher`.

Like `TimeLimitingCollector` and `ExitableDirectoryReader`, `TimeLimitingBulkScorer` is a low-level implementation detail for timeout support, that I think of as something that should remain private. I can be convinced to make it public but I would like to understand what are the use-cases that need to leverage this `TimeLimitingBulkScorer` directly as opposed to leveraging `IndexSearcher`'s built-in timeout support.

**Comment by reta** [comment]
Thanks @jpountz , 

> I can be convinced to make it public but I would like to understand what are the use-cases that need to leverage this TimeLimitingBulkScorer directly as opposed to leveraging IndexSearcher's built-in timeout support.

I think I presented one here (and actually Elasticsearch has the same [1] unsurprisingly): the `IndexSearcher` methods could be overridden (multitude of reasons, I have no credits here), including `IndexSearcher::search`, but the timeout support (BulkScorer etc) is sealed and not reusable.  

[1] https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java#L166

**Comment by reta** [comment]
> Then it should be possible to use the existing `IndexSearcher#search` logic as-is?

Thanks @jpountz, I think that would be the best option for everyone (Lucene / OpenSearch / Elasticsearch), but it would certainly fall into "high risk" change without reverse engineering and matching both implementations (I don't have full context sadly). I am closing this pull request (and related issue), we may get back to it at some point in the future, but for the time being it could not be used in OpenSearch. Thank you for your time.

**Pull request closed without merging**